### PR TITLE
openssl: add 3.0.8 + add GitHub as mirror

### PR DESCRIPTION
**Pull request by Croydon**
https://www.openssl.org/news/openssl-3.0-notes.html

For 1.1.1t see #15860

___

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by RubenRBS — DISMISSED** [review]

**Review by prince-chrismc — DISMISSED** [review]

**Comment by gegles** [comment]
What's the status on this? Is anybody actively looking into it?

Much appreciated!

**Comment by prince-chrismc** [comment]
> What's the status on this? Is anybody actively looking into it?

You can checkout this bot https://github.com/prince-chrismc/conan-center-index-pending-review/issues/1 which will give you a rundown of all the PRs and the bottom has graphs about how long they are taking

**Comment by prince-chrismc** [comment]
Ill need to check why this failed again from my MacBook 😕

**Comment by Croydon** [comment]
> Ill need to check why this failed again from my MacBook 😕

Is this something that happened recently? Can you give me a cross-reference please?

This PR doesn't modify 3.0.5 at all, so it is weird.

**Comment by gegles** [comment]
Any movement on this? We're waiting on this to release our product so we can incorporate the latest OpenSSL security fixes...

**Comment by Croydon** [comment]
Random thing that came up with a search: Maybe the openssl config files is making problems while loading the legacy stuff https://github.com/nodejs/node/discussions/43184

The question remains what changed in the CI environment that surfaced this error? It is really bad that critical updates like these are blocked by unrelated bugs. 

Maybe we should disable the tests with the legacy providers to get this merged first.

**Comment by clauverjat** [comment]
> Random thing that came up with a search: Maybe the openssl config files is making problems while loading the legacy stuff [nodejs/node#43184](https://github.com/nodejs/node/discussions/43184)
> 
> The question remains what changed in the CI environment that surfaced this error? It is really bad that critical updates like these are blocked by unrelated bugs.
> 
> Maybe we should disable the tests with the legacy providers to get this merged first.

I am following this issue closely as we are waiting on this PR being merged to make a new release of our software to respond to the disclosed openssl vulnerabilities. I could not agree more with your last statement.

Hope this issue will be resolved soon.

Thanks for your work !

**Comment by Croydon** [comment]
We are actually already skipping tests for OpenSSL 1.x on macOS https://github.com/conan-io/conan-center-index/blob/6a19ce1805fa4bf185151f66eeed4da0abdab067/recipes/openssl/1.x.x/test_v1_package/conanfile.py#L12

**Comment by SpaceIm** [comment]
> We are actually already skipping tests for OpenSSL 1.x on macOS armv8 shared
> 
> https://github.com/conan-io/conan-center-index/blob/6a19ce1805fa4bf185151f66eeed4da0abdab067/recipes/openssl/1.x.x/test_v1_package/conanfile.py#L12

yes but for another reason (and this skip_test in 1.x.x branch should be removed because CMake has been upgraded on Macos agents).

**Comment by Croydon** [comment]
I split out 1.1.1t to get it merged quicker: #15842

3.x will stay in this PR

**Comment by SpaceIm** [comment]
About this error: why openssl recipe doesn't define `OPENSSL_ENGINES` & `OPENSSL_MODULES` env vars in `package_info()`?

**Review by exief — APPROVED** [review]

**Comment by gegles** [comment]
Any hope this will get merged today?

Thanks for all the hard work!

**Comment by prince-chrismc** [comment]
> ```
> Loading legacy provider failed with this error:
> 	error:12800067:DSO support routines::could not load the shared library
> ```
> 
> Why openssl recipe doesn't define `OPENSSL_ENGINES` & `OPENSSL_MODULES` env vars in `package_info()`?

This error is blocking all the 3.x PRs ... must be something showing with client fixes

**Review by exief — DISMISSED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by danisevas** [comment]
any updates on this one?

**Comment by prince-chrismc** [comment]
@danisevas feel free to contribute a fix to help get this to pass :)

**Comment by SpaceIm** [comment]
> ```
> Loading legacy provider failed with this error:
> 	error:12800067:DSO support routines::could not load the shared library
> ```
> 
> Why openssl recipe doesn't define `OPENSSL_ENGINES` & `OPENSSL_MODULES` env vars in `package_info()`?

Reminder

**Comment by prince-chrismc** [comment]
Do that plus add some links for reviewers and it can go in :)

**Comment by SpaceIm** [comment]
It's a suggestion which would make sense given the error and OpenSSL documentation regarding env var and loading mechanism of modules. I let OpenSSL experts here give their opinion and provide the fix.

**Comment by daniel-heater-imprivata** [comment]
Can this be replicated locally?

I tried it on my Mac with
```
[settings]
arch=x86_64
build_type=Debug
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=14
os=Macos
[options]
openssl:shared=False
```

Both `conan create . openssl/3.0.8@ --build=missing` and `conan test test_package openssl/3.0.8@` succeed.

**Comment by gegles** [comment]
@prince-chrismc @SpaceIm anything that can be done to get this merged faster? We're holding off an entire release because of this. 

Much appreciated!

**Comment by prince-chrismc** [comment]
@gegles please contribute the fix 🙏 that will ve the fastest way :)

**Comment by daniel-heater-imprivata** [comment]
I have not been able to reproduce locally and I do not have access to the full CI logs. What is the best way to contribute?

Also, I diffed the 3.0.7 and 3.0.8 release tags and this changes looks like it might be relevant:
```
diff --git a/crypto/md5/build.info b/crypto/md5/build.info
index a1e28c8153..9a32538606 100644
--- a/crypto/md5/build.info
+++ b/crypto/md5/build.info
@@ -23,7 +23,7 @@ SOURCE[../../libcrypto]=$COMMON
 # default provider.  A no-deprecated build removes the external definition from
 # libcrypto and this means that the code needs to be in liblegacy.  However,
 # when building without 'dso', liblegacy is included in libcrypto.
-IF[{- !$disabled{dso} -}]
+IF[{- !$disabled{module} && !$disabled{shared} -}]
   SOURCE[../../providers/liblegacy.a]=$COMMON
 ENDIF
```

**Comment by prince-chrismc** [comment]
You can access the logs from 
the comments of the bot ... the last one is https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/15802/10-configs/macos-clang/openssl/3.0.8//summary.json

You might need to try using the docker images https://github.com/conan-io/conan-center-index/blob/master/docs/developing_recipes_locally.md#testing-more-environments since there is most likely going to be env dependant

**Comment by daniel-heater-imprivata** [comment]
> EDIT: I suggested docker images but that wont work for macos nodes

[osxcross](https://github.com/tpoechtrager/osxcross)?

**Comment by jcar87** [comment]
Hi @Croydon - we are currently investigating this.

Could you please add the following to the recipe so that we can verify the build configuration? I've changed the PR to draft while we continue to investigate this

````diff
diff --git a/recipes/openssl/3.x.x/conanfile.py b/recipes/openssl/3.x.x/conanfile.py
index 05de257a5..2a7b19e88 100644
--- a/recipes/openssl/3.x.x/conanfile.py
+++ b/recipes/openssl/3.x.x/conanfile.py
@@ -610,6 +610,7 @@ class OpenSSLConan(ConanFile):
                 self._create_targets()
                 with self._make_context():
                     self._make()
+                    self.run("perl source_subfolder/configdata.pm --dump")
 
     @property
     def _win_bash(self)
````

**Comment by jcar87** [comment]
There are two "loadable" modules in the package, irrespective of whether the recipe is built as `shared` or not:
```
├── lib
│   ├── cmake
│   │   └── conan-official-openssl-variables.cmake
│   ├── libcrypto.a
│   ├── libssl.a
│   └── ossl-modules
│       ├── fips.dylib
│       └── legacy.dylib
└── licenses
```

The issue is that OpenSSL will look in the directory provided at install time - this is why we cannot reproduce this locally, because those directories will exist on everyone's local machine as part of Conan create.

An easier way to test this is to clear the binaries from the Conan cache, and run:
```
conan test . openssl/3.0.7@ 
```

and let Conan fetch the OpenSSL binaries from Conan Center - we should the same error. Inspecting the process, I can see it tries to look in a path inside `/Users/jenkins`, which doesn't exist on my machine, and this is likely what's heppening in those failed builds above.

Adding this to the `package_info` seems to fix it for me:
```diff
diff --git a/recipes/openssl/3.x.x/conanfile.py b/recipes/openssl/3.x.x/conanfile.py
index 05de257a5..34311f48a 100644
--- a/recipes/openssl/3.x.x/conanfile.py
+++ b/recipes/openssl/3.x.x/conanfile.py
@@ -756,3 +756,9 @@ class OpenSSLConan(ConanFile):
         self.cpp_info.components["crypto"].names["cmake_find_package_multi"] = "Crypto"
         self.cpp_info.components["ssl"].names["cmake_find_package"] = "SSL"
         self.cpp_info.components["ssl"].names["cmake_find_package_multi"] = "SSL"
+
+        openssl_modules_dir = os.path.join(self.package_folder, "lib", "ossl-modules")
+        self.runenv_info.define_path("OPENSSL_MODULES", openssl_modules_dir)
+
+        # For legacy 1.x downstream consumers, remove once recipe is 2.0 only:
+        self.env_info.OPENSSL_MODULES = openssl_modules_dir
```

This is a rough patch that should fix the issue - I'd advise to also add some comments in the recipe clarifying the need for this. THis env-var is documented in: https://www.openssl.org/docs/man3.0/man7/openssl-env.html

Thanks!

**Comment by uilianries** [comment]
@Croydon I just pushed @jcar87 suggestions to this branch, please, don't forget to rebase on your local copy.

File: recipes/openssl/3.x.x/conanfile.py [hunk 1]
```diff
@@ -610,6 +610,7 @@ def build(self):
                 self._create_targets()
                 with self._make_context():
                     self._make()
+                    self.run("perl source_subfolder/configdata.pm --dump")
```
Comment (SpaceIm) [inline recipes/openssl/3.x.x/conanfile.py:613]: what is it?

Comment (SSE4) [reply recipes/openssl/3.x.x/conanfile.py:613]: yes, it's code to dump OpenSSL configuration

File: recipes/openssl/3.x.x/conanfile.py [hunk 2]
```diff
@@ -756,3 +757,10 @@ def package_info(self):
         self.cpp_info.components["crypto"].names["cmake_find_package_multi"] = "Crypto"
         self.cpp_info.components["ssl"].names["cmake_find_package"] = "SSL"
         self.cpp_info.components["ssl"].names["cmake_find_package_multi"] = "SSL"
+
+        openssl_modules_dir = os.path.join(self.package_folder, "lib", "ossl-modules")
+        self.runenv_info.define_path("OPENSSL_MODULES", openssl_modules_dir)
+
+        # For legacy 1.x downstream consumers, remove once recipe is 2.0 only:
+        self.env_info.OPENSSL_MODULES = openssl_modules_dir
```
Comment (SSE4) [inline recipes/openssl/3.x.x/conanfile.py:765]: can we set more variables listed here? https://www.openssl.org/docs/man3.0/man7/openssl-env.html

File: recipes/openssl/3.x.x/conanfile.py (hunk 1 above)
Comment (SSE4) [reply recipes/openssl/3.x.x/conanfile.py:613]: check this out: https://github.com/openssl/openssl/blob/master/Configure#L32-L48

Comment (prince-chrismc) [reply recipes/openssl/3.x.x/conanfile.py:613]: It was debug code from us testing it

Comment (SSE4) [reply recipes/openssl/3.x.x/conanfile.py:613]: it might be useful for troubleshooting in future, e.g. debugging bugs, it's worth keeping

**Comment by Croydon** [comment]
Thanks everyone for investigating this.

The CI's main job failed without a reported error. I'm going to restart the CI

**Review by exief — APPROVED** [review]

**Review by jcar87 — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Pull request merged**